Basic UI and graphics

aruu opened the pull request:
Please critique anywhere where the commenting or coding style could be better

aruu committed:
Trying out <canvas>

aruu committed:
Fix up indentation

aruu committed:
Basic UI layout and grid display

aruu committed:
Fix main.js from Windows to Unix line-endings

aruu committed:
Janky hiding system (old)

aruu committed:
Using CSS Grid for positioning

aruu committed:
Hiding overlay elements by just moving them instead of using CSS transitions

aruu committed:
Preparing for code review
Good except for hiding mechanism

aruu committed:
Adding HTML meta tag

aruu committed:
Clean up Toolbar Hide animation, comment

aruu commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Basic frontend UI](https://trello.com/c/lPIFozUo/3-basic-frontend-ui)

dmokracy commented on js/main.js:
```diff
@@ -0,0 +1,76 @@
+
+var canvas = document.querySelector(".canvas");
+var ctx = canvas.getContext("2d");
+ctx.canvas.height = window.innerHeight - 4;  // slight offset or else it overflows and causes scroll in Firefox
```
Any way to avoid hard-coding this offset?

aruu replied:
...found some web standards/DocType stuff that is way over my head
https://stackoverflow.com/a/12094962
https://www.khronos.org/webgl/public-mailing-list/public_webgl/1011/msg00097.php
But the fix works so I'll remove the hard-coded offset

aruu replied:
Addressed in 2fe5049

dmokracy commented on index.html:
```diff
@@ -1,11 +1,25 @@
+        <meta charset="utf-8">
+        <title>Conway's Game of Life</title>
+        <!-- Normalize.css to reset styles -->
+        <link rel="stylesheet" href="https://unpkg.com/normalize.css@8.0.0/normalize.css">
```
It looks like best practice is to use subresource integrity for any third-party resources, like what was done for the Font Awesome CSS below.
https://www.srihash.org/

aruu replied:
Good catch, both were just HTML I copied from their sites. Never seriously considered this in real-life, even though I recently read a blog post on it looooool. Time to put knowledge into practice.
https://dev.to/dotnetcoreblog/three-steps-for-increasing-the-security-of-your-web-apps-3clg

dmokracy commented on css/main.css:
```diff
@@ -0,0 +1,50 @@
+        "   .       .      tool  "
+        "   .       .      tool  "
+        "control    .        .   ";
+}
```
Warning: grid isn't supported on Internet Explorer!!!
Oh wait nvm no one (not even Github) cares about supporting IE, so this is fine.

dmokracy commented on js/main.js:
```diff
@@ -0,0 +1,76 @@
+                e.style[hideIterator[dir].style] = e[hideIterator[dir].size] + "px";
+            } else {
+                e.style[hideIterator[dir].style] = "0px";
+            }
```
(Suggestion Only)
Could probably express this if-statement more compactly with a ternary operator.

dmokracy commented on js/main.js:
```diff
@@ -0,0 +1,76 @@
+    let hideBtnIconClassList = hideBtn.querySelector("i").classList;
+    hideBtnIconClassList.toggle("fa-angle-double-up");
+    hideBtnIconClassList.toggle("fa-angle-double-down");
+    let isHidden = hideBtnIconClassList.contains("fa-angle-double-down");
```
(Suggestion Only)
This should also work:
let isHidden = hideBtnIconClassList.toggle("fa-angle-double-down");
But your implementation seems more readable to me.

dmokracy commented on js/main.js:
```diff
@@ -0,0 +1,76 @@
+var pixelSize = 25;
+var borderSize = 2;      // Border is drawn on the right and bottom edges of the pixel
+ctx.fillStyle = "gray";  // This will be the border colour
+ctx.fillRect(0, 0, ctx.canvas.width, ctx.canvas.height);
```
(Subjective) 
I've generally been taught to put comments on the line above code, rather than on the same line.

aruu replied:
Comments reorganized, yeah it's more readable when above for these

aruu committed:
Remove hard-coded height-offset to fit <canvas>

Refer to
https://stackoverflow.com/a/12094962
https://www.khronos.org/webgl/public-mailing-list/public_webgl/1011/msg00097.php

aruu committed:
Use SRI for normalize.css

aruu committed:
Refactor toolbar hiding code

- rename "hideIterator" to "hideLogicObj"
- use an actual status class on the button for status, rather than using the appearance as the indicator
- use a ternary operator one-liner for conditional assignment

aruu committed:
More readable comments hopefully

aruu commented:
Ready for review

dmokracy reviewed (approved):
Ship it!
